sc-memory/sc-memory/sc_agent.tpp

```diff
@@ -0,0 +1,580 @@
+
+  return it5->Get(0);
+}
+
+template <class TScEvent, class TScContext>
+ScAddr ScAgentAbstract<TScEvent, TScContext>::GetEventClass() const
+{
+  if constexpr (!std::is_same<TScEvent, ScElementaryEvent>::value)
+    return TScEvent::eventClassAddr;
+
+  ScIterator5Ptr const it5 = m_memoryCtx.Iterator5(
+      GetAbstractAgent(),
+      ScType::EdgeDCommonConst,
+      ScType::EdgeDCommonConst,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_primary_initiation_condition);
+  if (!it5->Next())
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidState,
+        "Abstract sc-agent `"
+            << this->GetName()
+            << "` has not primary initiation condition and result. Check that abstract sc-agent has specified "
+               "relation `nrel_primary_initiation_condition`.");
+
+  return m_memoryCtx.GetEdgeSource(it5->Get(2));
+}
+
+template <class TScEvent, class TScContext>
+ScAddr ScAgentAbstract<TScEvent, TScContext>::GetEventSubscriptionElement() const
+{
+  ScIterator5Ptr const it5 = m_memoryCtx.Iterator5(
+      GetAbstractAgent(),
+      ScType::EdgeDCommonConst,
+      ScType::EdgeDCommonConst,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_primary_initiation_condition);
+  if (!it5->Next())
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidState,
+        "Abstract sc-agent `"
+            << this->GetName()
+            << "` has not primary initiation condition and result. Check that abstract sc-agent has specified "
+               "relation `nrel_primary_initiation_condition`.");
+
+  return m_memoryCtx.GetEdgeTarget(it5->Get(2));
+}
+
+template <class TScEvent, class TScContext>
+ScAddr ScAgentAbstract<TScEvent, TScContext>::GetActionClass() const
+{
+  ScIterator5Ptr const it5 = m_memoryCtx.Iterator5(
+      GetAbstractAgent(),
+      ScType::EdgeDCommonConst,
+      ScType::NodeConstClass,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_sc_agent_action_class);
+  if (!it5->Next())
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidState,
+        "Abstract sc-agent `" << this->GetName()
+                              << "` has not action class. Check that abstract sc-agent has specified "
+                                 "relation `nrel_sc_agent_action_class`.");
+
+  return it5->Get(2);
+}
+
+// LCOV_EXCL_START
+template <class TScEvent, class TScContext>
+sc_bool ScAgentAbstract<TScEvent, TScContext>::CheckInitiationCondition(TScEvent const & event)
+{
+  return SC_TRUE;
+}
+
+// LCOV_EXCL_STOP
+
+template <class TScEvent, class TScContext>
+ScAddr ScAgentAbstract<TScEvent, TScContext>::GetInitiationCondition() const
+{
+  ScIterator5Ptr const it5 = m_memoryCtx.Iterator5(
+      GetAbstractAgent(),
+      ScType::EdgeDCommonConst,
+      ScType::EdgeDCommonConst,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_initiation_condition_and_result);
+  if (!it5->Next())
```

Comment:
This method is implemented in a library that depends on sc-memory lib - sc-kpm. The utilities in sc-kpm are a dump. It will be removed from the machine